prover/prover.go

```diff
@@ -191,47 +192,74 @@ func makeRecoveryProofTree(
 	return treeCache, tree, nil
 }
 
-func generateProof(
+func sequentialWork(
 	ctx context.Context,
 	labelHashFunc func(data []byte) []byte,
 	tree *merkle.Tree,
 	treeCache *cache.Writer,
-	end time.Time,
 	nextLeafID uint64,
 	securityParam uint8,
 	persist persistFunc,
-) (uint64, *shared.MerkleProof, error) {
-	makeLabel := shared.MakeLabelFunc()
+) (uint64, error) {
 	leaves := nextLeafID
-	for leafID := nextLeafID; time.Until(end) > 0; leafID++ {
-		// Handle persistence.
+	makeLabel := shared.MakeLabelFunc()
+	for leafID := nextLeafID; ; leafID++ {
 		select {
 		case <-ctx.Done():
 			if err := persist(ctx, tree, treeCache, leafID); err != nil {
-				return 0, nil, fmt.Errorf("%w: error happened during persisting: %v", ErrShutdownRequested, err)
+				return 0, fmt.Errorf("persisting execution state: %w", err)
 			}
-			return 0, nil, ErrShutdownRequested
+			return leaves, nil
```

Comment:
imo when returning early due to `ctx.Done()` the returned error should be `ctx.Err()`

Reply:
The thing is that this condition is not "returning early". This is the main return point meaning "the work was done". When the `ctx` in `sequentialWork` is cancelled it means "stop producing more leaves". It was achieved via `time.Until(end) <= 0` before. Now the `ctx` is cancelled instead.